### PR TITLE
AddPlasmaFromFile fix for periodic boundaries

### LINKED ISSUE #1088: openPMD with periodic boundary conditions
Hi all,

When reading particles that fall outside of the simulation box, using openPMD external input file via `<species>.injection_style = external_file` and periodic boundary conditions, those particles are reflected back into the box (top left plot).

image

Above are plots from my usual openPMD particle files.
I will try to see how best to fix this and share all updates here.

Cheers,
Diana

### PR DESCRIPTION
**Pull request by LDAmorim**
This PR replaces the 
```
if (plasma_injector->insideBounds(x, y, z))
```
condition inside the `PhysicalParticleContainer::AddPlasmaFromFile` function of the file PhysicalParticleContainer.cpp, by `if` statements that check that the particles are within the simulation box bounds.

This fixed issue #1088 (bottom right plot, as opposed to top left) where an issue was occurring for periodic boundary conditions, if the particles in the file were off the simulation box.

image

I repeated the test in 2D (but with my 3D openPMD input files), and it also seems to work well with this fix, when using periodic boundary conditions.

### CODE REVIEW

**Comment by LDAmorim** [comment]
Thanks for checking this PR @ax3l.
ps: I think I had already linked this to the issue.

**Comment by ax3l** [comment]
Yes, thank you for that!

You can use magic syntax such as `close #...` or `fix #...` etc. if you want to auto-close an issue with a PR :) https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

File: Source/Particles/PhysicalParticleContainer.cpp [hunk 1]
```diff
@@ -358,27 +358,49 @@ PhysicalParticleContainer::AddPlasmaFromFile(ParticleReal q_tot,
             weight = w * w_unit;
         }
 
+        // TODO: Add option for different levels of injection
```
Comment (ax3l) [inline Source/Particles/PhysicalParticleContainer.cpp]: Can you please help me what this refers to? I am just confused because we inject/run/store particles always on the finest level of the refined mesh. I assume this currently injects particles always to the coarsest grid?

File: Source/Particles/PhysicalParticleContainer.cpp [hunk 2]
```diff
@@ -358,27 +358,49 @@ PhysicalParticleContainer::AddPlasmaFromFile(ParticleReal q_tot,
...
+        const auto probhi = geom.ProbHiArray();
+
         for (auto i = decltype(npart){0}; i<npart; ++i){
             ParticleReal const x = ptr_x.get()[i]*position_unit_x;
-            ParticleReal const z = ptr_z.get()[i]*position_unit_z+z_shift;
+            if (x>problo[0] && x<probhi[0]){
+                ParticleReal const z = ptr_z.get()[i]*position_unit_z+z_shift;
 #   ifndef WARPX_DIM_3D
-            ParticleReal const y = 0.0_prt;
+                if (z>problo[1] && z<probhi[1]){
```
Comment (ax3l) [inline Source/Particles/PhysicalParticleContainer.cpp]: I am not sure the if cascased here are logically stable. A particle should only be initialized if it satisfies all contrains (x,y,z), so maybe they statements in the `if(...)`s should be `&&` concatenated, stored in a temporary boolean variable such as `is_valid` and then we initialize it?

Comment (LDAmorim) [reply Source/Particles/PhysicalParticleContainer.cpp:371]: The problem is in 2D runs, this won't work unless you set the y to 0.

File: Source/Particles/PhysicalParticleContainer.cpp (hunk 1 above)
Comment (LDAmorim) [reply Source/Particles/PhysicalParticleContainer.cpp]: I am not sure of the future, it is a possible hypothesis, so I wanted to leave the TODO in case anyone else would need to update it later.

File: Source/Particles/PhysicalParticleContainer.cpp (hunk 2 above)
Comment (LDAmorim) [reply Source/Particles/PhysicalParticleContainer.cpp]: You can check the beam definition above this function as an example.

File: Source/Particles/PhysicalParticleContainer.cpp (hunk 1 above)
Comment (LDAmorim) [reply Source/Particles/PhysicalParticleContainer.cpp]: Sorry, I missed a part of your question. Yes, this only injects to the coarsest grid. But later we might want to inject to other grids.

File: Source/Particles/PhysicalParticleContainer.cpp (hunk 2 above)
Comment (LDAmorim) [reply Source/Particles/PhysicalParticleContainer.cpp]: I am not sure I understood your point here very clearly.
As the code is, the particle is only initialized if it meets all the if constrains.
With the cascade option, if the particle fails to be inside in the x dimension, we don't loose time to check all others.

File: Source/Particles/PhysicalParticleContainer.cpp (hunk 1 above)
Comment (ax3l) [reply Source/Particles/PhysicalParticleContainer.cpp]: Okay, the comment was just not clear to me. Maybe
```suggestion
        // TODO: This currently only injects particles on the coarsest grid
```
is more clear?

File: Source/Particles/PhysicalParticleContainer.cpp (hunk 2 above)
Comment (ax3l) [reply Source/Particles/PhysicalParticleContainer.cpp]: I think the nested logic makes it hard to read this part. I suggest to create and intermediate variable instead of nesting multiple ifs inside a for loop. Let me add a code suggestion.

Comment (LDAmorim) [reply Source/Particles/PhysicalParticleContainer.cpp]: Oh ok. Don't worry! I can try to do it.

Comment (ax3l) [reply Source/Particles/PhysicalParticleContainer.cpp]: I opened the file locally now and understand what you did: you try to avoid loading data when you already know the particle will not be initialized anyway. I feel this is probably an optimization that is not super critical, but if you like to add it maybe use blocks like these:
```C++
            bool const in_box_x = x>problo[0] && x<probhi[0];
            if( not in_x_bounds ) continue;
```
Instead of nesting the `ifs`. This names the constraints that we check here and keeps the structure flat. Otherwise I feel this gets unnecessarily hard to read, especially with the interleaved preprocessor commands.

File: Source/Particles/PhysicalParticleContainer.cpp (hunk 1 above)
Comment (LDAmorim) [reply Source/Particles/PhysicalParticleContainer.cpp]: Hum, you are right. It was not clear! But yours is not really a todo... What do you think of the new option?

File: Source/Particles/PhysicalParticleContainer.cpp (hunk 2 above)
Comment (LDAmorim) [reply Source/Particles/PhysicalParticleContainer.cpp]: That sounds good! I will do that and test it, thanks @ax3l !

File: Source/Particles/PhysicalParticleContainer.cpp (hunk 1 above)
Comment (ax3l) [reply Source/Particles/PhysicalParticleContainer.cpp]: `Note: `/`Fixme: `? :D

Comment (LDAmorim) [reply Source/Particles/PhysicalParticleContainer.cpp]: Great idea!!

**Comment by LDAmorim** [comment]
Tested last changes with same input (w periodic boundary conditions) and it seems to work well in 3D:
image

**Comment by LDAmorim** [comment]
I double checked all the other properties of the beams and everything seems to be matching well:

image

And I get the expected WarpX warning message in the output:
```bash
WARNING: Simulation box doesn't cover all particles
```

**Comment by LDAmorim** [comment]
Test in 2D seems to not work well:
image

File: Source/Particles/PhysicalParticleContainer.cpp [hunk 3]
```diff
@@ -358,27 +358,38 @@ PhysicalParticleContainer::AddPlasmaFromFile(ParticleReal q_tot,
...
+            ParticleReal const uy = 0.0_prt;
 #   else
+            bool const in_box_z = z>problo[2] && z<probhi[2];
+            if( not in_box_z ) continue;
             ParticleReal const y = ptr_y.get()[i]*position_unit_y;
+            bool const in_box_y = y>problo[1] && y<probhi[1];
+            if( not in_box_y ) continue;
+            ParticleReal const uy = ptr_uy.get()[i]*momentum_unit_y/PhysConst::m_e;
 #   endif
-            if (plasma_injector->insideBounds(x, y, z)) {
```
Comment (ax3l) [inline Source/Particles/PhysicalParticleContainer.cpp]: I have the feeling that the optimization of reads is not too relevant and we could just try to fix this call here.

In fact, it looks like these bounds are set to inf by default (instead of the sim box) and that user-parameters are already present:
https://github.com/ECP-WarpX/WarpX/blob/master/Source/Initialization/PlasmaInjector.cpp#L91-L96

**Comment by LDAmorim** [comment]
Update on the 2D tests:
The issue mentioned in my previous comment was due to my analysis.
I had forgotten to remove the selection in y to the openPMD input data files and to not account for vy (as in WarpX, we are setting it to 0).

Now the # of particles, their position and velocities agree in all runs (with and without periodic boundaries) as can be seen in the top raow of the plots below:

image

File: Source/Particles/PhysicalParticleContainer.cpp (hunk 3 above)
Comment (LDAmorim) [reply Source/Particles/PhysicalParticleContainer.cpp]: Thanks, will change `plasma_injector` default to box so we can make the fix of the current PR applicable to all the injection schemes.
And will update documentation on this too.

**Comment by LDAmorim** [comment]
Tested current commits in 3D and results (using Periodic boundary conditions) match to expected values:
image

**Comment by LDAmorim** [comment]
Also the 2D test leads to correct results:
image

File: Docs/source/running_cpp/parameters.rst [hunk 4]
```diff
@@ -288,6 +288,12 @@ Particle initialization
     The mass of one `physical` particle of this species.
     If ``species_type`` is specified, the mass will be set to the physical value and ``mass`` is optional.
 
+* ``<species_name>.xmin,ymin,zmin`` (`float`) optional (default `geometry.prob_lo[0],[1],[2]`)
```
Comment (ax3l) [inline Docs/source/running_cpp/parameters.rst]: I think the default is not the simulation box (`prob_lo/hi`) as we saw in the implementation - but essentially `-inf/inf`:
https://github.com/ECP-WarpX/WarpX/blob/787bf0631809f6929bcd962cb8954063141e6490/Source/Initialization/PlasmaInjector.cpp#L82-L89

cc @atmyers

Comment (LDAmorim) [reply Docs/source/running_cpp/parameters.rst]: Which gives the error when using the periodic boundary conditions.

Comment (LDAmorim) [reply Docs/source/running_cpp/parameters.rst]: But I fixed the min, max to the box boundaries in this PR, that is why I updated the Docs with that information.

Comment (ax3l) [reply Docs/source/running_cpp/parameters.rst]: ha, scrolling helps. sorry for that!

File: Source/Initialization/PlasmaInjector.cpp [hunk 5]
```diff
@@ -79,14 +79,22 @@ PlasmaInjector::PlasmaInjector (int ispecies, const std::string& name)
...
-    // parse plasma boundaries
-    xmin = std::numeric_limits<amrex::Real>::lowest();
-    ymin = std::numeric_limits<amrex::Real>::lowest();
-    zmin = std::numeric_limits<amrex::Real>::lowest();
-
-    xmax = std::numeric_limits<amrex::Real>::max();
-    ymax = std::numeric_limits<amrex::Real>::max();
-    zmax = std::numeric_limits<amrex::Real>::max();
+    // box boundaries
+    // FIXME: Only injects particles on coarsest grid, so changes are needed to allow injection to finer grid levels.
```
Comment (ax3l) [inline Source/Initialization/PlasmaInjector.cpp]: the fixme looks confusing to me at this point, where we define the geometry. It's more of a general comment for the whole plasma injector class I think.
In the annotated block, we just define the xmin/xmax et al., which are always the "widest" grid.

File: Source/Initialization/PlasmaInjector.cpp [hunk 6]
```diff
@@ -79,14 +79,22 @@ PlasmaInjector::PlasmaInjector (int ispecies, const std::string& name)
...
+    const Geometry& geom = WarpX::GetInstance().Geom(0);
```
Comment (ax3l) [inline Source/Initialization/PlasmaInjector.cpp:92]: Object lifetime question: is the `PlasmaInjector` class created once at the simulation start, which will probably return the t=0 sim box extend here, or again every time step? If the former, then we might cause problems fixing it to the t=0 geometry with continuous injection in moving windows, e.g. for brackground plasma.

File: Source/Initialization/PlasmaInjector.cpp [hunk 7]
```diff
@@ -497,8 +505,12 @@ XDim3 PlasmaInjector::getMomentum (Real x, Real y, Real z) const noexcept
 bool PlasmaInjector::insideBounds (Real x, Real y, Real z) const noexcept
 {
     return (x < xmax and x >= xmin and
+#ifndef WARPX_DIM_3D
```
Comment (ax3l) [inline Source/Initialization/PlasmaInjector.cpp]: this will disable the y bounds check for particles in RZ. A better check might be ([docs](https://warpx.readthedocs.io/en/latest/developers/dimensionality.html)):
```suggestion
#ifdef WARPX_DIM_XZ
```

Not using the `y` parameter will cause a compile-time warning in 2D. We can add `(void)y;` before the return line to suppress it.

Potentially cleaner is to not modify the `insideBounds` function at all and instead keep in `WARPX_DIM_XZ` the `ymin`/`ymax` with `std::numeric_limits<amrex::Real>::lowest()`/`std::numeric_limits<amrex::Real>::max()` instead, see note above.

File: Source/Initialization/PlasmaInjector.cpp [hunk 8]
```diff
@@ -79,14 +79,22 @@ PlasmaInjector::PlasmaInjector (int ispecies, const std::string& name)
...
+    xmax = probhi[0];
+#ifndef WARPX_DIM_3D
+    zmin = problo[1];
+    zmax = probhi[1];
+#else
+    ymin = problo[1];
+    ymax = probhi[1];
+    zmin = problo[2];
+    zmax = probhi[2];
+#endif
```
Comment (ax3l) [inline Source/Initialization/PlasmaInjector.cpp]: This will disable keep the some variables uniitialized in 2D, which can lead to segfaults and many `#if` needs down the road. For example, it also disables bounds check for particles in RZ ([docs](https://warpx.readthedocs.io/en/latest/developers/dimensionality.html)) or could lead to segfaults with new geometries.

I would suggest the following:
```suggestion
    ymin = problo[1];
    ymax = probhi[1];
    zmin = problo[2];
    zmax = probhi[2];

#ifdef WARPX_DIM_XZ
    ymin = std::numeric_limits<amrex::Real>::lowest();
    ymax = std::numeric_limits<amrex::Real>::max();
    zmin = problo[1];
    zmax = probhi[1];
#endif
```

**Comment by LDAmorim** [comment]
It is not clear how the changes to the defaults of the plasma injection Bounds function can affect particles using `do_continuous_injection = 1`.
So I am testing this using `constant` and `parse_function` injection style plasmas with and without setting `xmin, xmax, etc`.

File: Source/Particles/PhysicalParticleContainer.cpp [hunk 9]
```diff
@@ -363,20 +363,18 @@ PhysicalParticleContainer::AddPlasmaFromFile(ParticleReal q_tot,
             ParticleReal const z = ptr_z.get()[i]*position_unit_z+z_shift;
 #   ifndef WARPX_DIM_3D
             ParticleReal const y = 0.0_prt;
+            ParticleReal const uy = 0.0_prt;
 #   else
             ParticleReal const y = ptr_y.get()[i]*position_unit_y;
+            ParticleReal const uy = ptr_uy.get()[i]*momentum_unit_y/PhysConst::m_e;
```
Comment (ax3l) [inline Source/Particles/PhysicalParticleContainer.cpp]: You now run into the risk that `uy` is not defined in a 2D input file and `ptr_uy` is a `nullptr`. I would just suggest to keep the logic that is below with
```C++
                ParticleReal uy = 0.0_prt;
                if (ps["momentum"].contains("y")) {
                    uy = ptr_uy.get()[i]*momentum_unit_y/PhysConst::m_e;
                }
```

and don't mix position and momentum component loading here.

Comment (LDAmorim) [reply Source/Particles/PhysicalParticleContainer.cpp]: I think you read the `#ifndef` wrong. Those 2 lines are read only when we run in 3D

Comment (LDAmorim) [reply Source/Particles/PhysicalParticleContainer.cpp]: In that case we assume that openPMD will crash the run if any component is missing (we say which ones are required in the Docs) and then we get the error message that the component is missing.

Comment (LDAmorim) [reply Source/Particles/PhysicalParticleContainer.cpp]: In 2D we set `y` and `uy` to 0.

File: Source/Initialization/PlasmaInjector.cpp (hunk 5 above)
Comment (LDAmorim) [reply Source/Initialization/PlasmaInjector.cpp]: Thanks Axel, that is a good point. How about?
```suggestion
    // NOTE: Default injection range is set to the dimensions of the mother grid.
```

File: Source/Initialization/PlasmaInjector.cpp (hunk 6 above)
Comment (LDAmorim) [reply Source/Initialization/PlasmaInjector.cpp:92]: That problem would already occur without these changes, when we define `zmin` and `zmax` values for plasmas that are injected from out of the box, or not?

File: Source/Initialization/PlasmaInjector.cpp (hunk 8 above)
Comment (LDAmorim) [reply Source/Initialization/PlasmaInjector.cpp]: Oh true that. Thanks for pointing that out!
I think then that it makes more sense to write it this way:
```suggestion
#ifndef WARPX_DIM_3D
    ymin = std::numeric_limits<amrex::Real>::lowest();
    ymax = std::numeric_limits<amrex::Real>::max();
    zmin = problo[1];
    zmax = probhi[1];
#else
    ymin = problo[1];
    ymax = probhi[1];
    zmin = problo[2];
    zmax = probhi[2];
#endif
```

File: Source/Initialization/PlasmaInjector.cpp (hunk 7 above)
Comment (LDAmorim) [reply Source/Initialization/PlasmaInjector.cpp]: I don't understand what you wrote here.
I am defining `y` and `uy` both. And I am using both in the `injectBounds()` function.
Also I tested in 2D and got no error, you can see the output in here:
[output.txt](https://github.com/ECP-WarpX/WarpX/files/4794569/output.txt)

Comment (LDAmorim) [reply Source/Initialization/PlasmaInjector.cpp]: Sorry, at compile time I don't remember seeing any warnings.
But I will check it again after confirming that plasmas initialized with moving window and outside of the box are injected correctly.

Comment (LDAmorim) [reply Source/Initialization/PlasmaInjector.cpp]: From the current version, I see no warnings at compile time.

File: Source/Initialization/PlasmaInjector.cpp (hunk 6 above)
Comment (LDAmorim) [reply Source/Initialization/PlasmaInjector.cpp:92]: Discussed offline: we will make an if that tests if moving window is being used. If that is the case then we set `zmin,zmax` to -/+inf

Comment (LDAmorim) [reply Source/Initialization/PlasmaInjector.cpp:92]: We decided to set zmin,zmax to -/+inf by default and only change it if in that direction periodic boundary conditions are being used.

**Comment by LDAmorim** [comment]
Last version was tested with a 3D run using periodic boundary conditions and introducing a plasma, with moving window and `do_continuous_injection=1`.
The results seem correct (I think a wider mirror would remove the fields seen behind the beam)!!

image

File: Source/Initialization/PlasmaInjector.cpp [hunk 10]
```diff
@@ -79,7 +79,7 @@ PlasmaInjector::PlasmaInjector (int ispecies, const std::string& name)
     pp.query("radially_weighted", radially_weighted);
     AMREX_ALWAYS_ASSERT_WITH_MESSAGE(radially_weighted, "ERROR: Only radially_weighted=true is supported");
 
-    // parse plasma boundaries
+    // box boundaries
```
Comment (LDAmorim) [inline Source/Initialization/PlasmaInjector.cpp]:
```suggestion
    // Unlimited boundaries
```

**Comment by LDAmorim** [comment]
2D test also gives correct results:
image

**Comment by LDAmorim** [comment]
Also in 2D the results seem correct:

image

File: Source/Initialization/PlasmaInjector.cpp [hunk 11]
```diff
@@ -88,6 +88,30 @@ PlasmaInjector::PlasmaInjector (int ispecies, const std::string& name)
     ymax = std::numeric_limits<amrex::Real>::max();
     zmax = std::numeric_limits<amrex::Real>::max();
 
+    // NOTE: When periodic boundaries are used, default injection range is set to mother grid dimensions.
+    const Geometry& geom = WarpX::GetInstance().Geom(0);
+    if( geom.isPeriodic(0)==1 ) {
```
Comment (ax3l) [inline Source/Initialization/PlasmaInjector.cpp]: Since this returns a `bool` we could make this more readability by
```suggestion
    if( geom.isPeriodic(0) == true ) {
```
or even
```suggestion
    if( geom.isPeriodic(0) ) {
```

File: Source/Initialization/PlasmaInjector.cpp [hunk 12]
```diff
@@ -88,6 +88,30 @@ PlasmaInjector::PlasmaInjector (int ispecies, const std::string& name)
...
+        xmin = geom.ProbLo(0);
+        xmax = geom.ProbHi(0);
+    }
+
+    if( geom.isPeriodic(1)==1 ) {
```
Comment (ax3l) [inline Source/Initialization/PlasmaInjector.cpp]:
```suggestion
    if( geom.isPeriodic(1) ) {
```

File: Source/Initialization/PlasmaInjector.cpp [hunk 13]
```diff
@@ -88,6 +88,30 @@ PlasmaInjector::PlasmaInjector (int ispecies, const std::string& name)
...
+        zmin = geom.ProbLo(1);
+        zmax = geom.ProbHi(1);
+#       else
+        ymin = geom.ProbLo(1);
+        ymax = geom.ProbHi(1);
+#       endif
+    }
+
+#   ifdef WARPX_DIM_3D
+    if( geom.isPeriodic(2)==1 ) {
```
Comment (ax3l) [inline Source/Initialization/PlasmaInjector.cpp]:
```suggestion
    if( geom.isPeriodic(2) ) {
```

**Review by ax3l — APPROVED** [review]
Thank you for spotting, fixing & testing this, Diana! :rocket: :sparkles:
